signac/synced_collections/utils.py

```diff
@@ -0,0 +1,219 @@
+    - JSON encoding of numpy arrays is not invertible; once encoded, reloading
+      the data will result in converting arrays to lists and numpy numbers into
+      ints or floats.
+    - This function assumes that the in-memory data for a SyncedCollection is
+      up-to-date. If the data has been changed on disk without updating the
+      collection, or if this function is used to serialize the data before any
+      method is invoked that would load the data from disk, the resulting
+      serialized data may be incorrect.
+
+    """
+    if NUMPY:
+        if isinstance(o, numpy.ndarray):
+            if o.shape == ():
+                return o.item()
+            else:
+                return o.tolist()
+        elif isinstance(o, numpy.number):
+            return o.item()
+    try:
+        return o._data
+    except AttributeError as e:
+        raise TypeError from e
+
+
+class SyncedCollectionJSONEncoder(JSONEncoder):
+    """A :class:`json.JSONEncoder` that handles objects encodeable using :func:`~.default`.
+
+    Warnings
+    --------
+    - JSON encoding of numpy arrays is not invertible; once encoded, reloading
+      the data will result in converting arrays to lists and numpy numbers into
+      ints or floats.
+    - This class assumes that the in-memory data for a SyncedCollection is
+      up-to-date. If the data has been changed on disk without updating the
+      collection, or if this class is used to serialize the data before any
+      method of the collection is invoked that would load the data from disk,
+      the resulting serialized data may be incorrect.
+
+    """
+
+    def default(self, o: Any) -> Dict[str, Any]:  # noqa: D102
+        try:
+            return default(o)
+        except TypeError:
+            # Call the super method, which raises a TypeError if it cannot
+            # encode the object.
+            return super().default(o)
+
+
+class _NullContext:
+    """A nullary context manager.
+
+    There are various cases where we sometimes want to perform a task within a
+    particular context, but at other times we wish to ignore that context. The
+    most obvious example is a lock for threading: since
+    :class:`~.SyncedCollection` allows multithreading support to be enabled or
+    disabled, it is important to be able to write code that is agnostic to
+    whether or not a mutex must be acquired prior to executing a task. Locks
+    support the context manager protocol and are used in that manner throughout
+    the code base, so the most transparent way to disable buffering is to
+    create a nullary context manager that can be placed as a drop-in
+    replacement for the lock so that all other code can handle this in a
+    transparent manner.
+    """
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def __call__(self):
+        """Allow usage of the context in a function-like manner."""
+        return self
```

Comment:
This is required because currently signac exposes the buffered context via `with collection.buffered()`, whereas the new version exposes it as `collection.buffered` (it's an attribute, not a method) and the same goes for "global" buffering. I did mean to ask whether we want to retain that backwards compatibility in signac 2.0; if not, we can mark this as deprecated.